src/Microsoft.Azure.WebJobs.Host/Executors/FunctionExecutor.cs

```diff
@@ -73,7 +73,59 @@ public HostOutputMessage HostOutputMessage
         public async Task<IDelayedException> TryExecuteAsync(IFunctionInstance functionInstance, CancellationToken cancellationToken)
         {
             var functionInstanceEx = (IFunctionInstanceEx)functionInstance;
-            var logger = _loggerFactory?.CreateLogger(LogCategories.CreateFunctionCategory(functionInstance.FunctionDescriptor.LogName));
+            var logger = _loggerFactory.CreateLogger(LogCategories.CreateFunctionCategory(functionInstanceEx.FunctionDescriptor.LogName));
+
+            if (functionInstance.FunctionDescriptor.RetryStrategy != null)
+            {
+                return await ExecuteWithRetriesAsync(functionInstanceEx, cancellationToken, logger, functionInstance.FunctionDescriptor.RetryStrategy);
+            }
+
+            return await TryExecuteAsyncCore(functionInstanceEx, cancellationToken, logger);
+        }
+
+        internal async Task<IDelayedException> ExecuteWithRetriesAsync(IFunctionInstanceEx functionInstance, CancellationToken token, ILogger logger, IRetryStrategy retryStrategy)
+        {
+            var attempt = 0;
+            IDelayedException functionResult = null;
+            TimeSpan nextDelay;
+            do
+            {
+                if (token.IsCancellationRequested)
+                {
+                    logger.LogDebug($"Function execution cancelled.");
+                    break;
+                }
+                logger.LogFunctionRetryAttempt(attempt, retryStrategy.MaxRetryCount);
+                functionResult = await TryExecuteAsyncCore(functionInstance, token, logger);
+                if (functionResult == null)
+                {
+                    // function invocation succeeded.
+                    break;
+                }
+
+                // Build retry context
+                var retryContext = new RetryContext
+                {
+                    RetryCount = ++attempt,
+                    Exception = functionResult.Exception,
+                    Instance = functionInstance
+                };
+
+                if (attempt >= retryStrategy.MaxRetryCount)
```

Comment:
nit: might move this above creation of the retry context, since if we're a limit, it won't be needed.